appsec: preview version

Julio-Guerra opened the pull request:
### What does this PR do?

Bootstrap a first version of the AppSec agent integrated into the trace agent as a reverse proxy.

### Describe how to test your changes

Usual HTTP server tests

Julio-Guerra committed:
appsec: bootstrap a first simple http server mux

Julio-Guerra committed:
appsec: bootstrap the batch aggregator

Julio-Guerra committed:
appsec/backend: first simple http client

Julio-Guerra committed:
appsec/api/http: http api config

Julio-Guerra committed:
appsec/agent: add the event pipeline

Julio-Guerra committed:
trace: integrate and configure the appsec agent via the trace agent

Julio-Guerra committed:
appsec: fix type errors

Julio-Guerra committed:
appsec/config: add appsec-specific configuration

Julio-Guerra committed:
appsec: fix the agent-enabled condition

Julio-Guerra committed:
appsec: periodically flush the batch

Julio-Guerra committed:
appsec: aligning with the latest RFC state

Integrate into the existing trace-agent HTTP server as a transparent
reverse proxy behind /appsec, redirected to the intake api url.

Julio-Guerra committed:
appsec: fix golint errors

Julio-Guerra committed:
Merge remote-tracking branch 'origin/main' into appsec-agent-beta

Julio-Guerra committed:
appsec/backend/proxy: remove unused const value

Julio-Guerra committed:
trace/api/info: add /appsec to the list of endpoints

Julio-Guerra committed:
appsec: code cleanup

Julio-Guerra committed:
Merge remote-tracking branch 'origin/main' into appsec-agent-beta

gbbr commented on pkg/config/config.go:
```diff
@@ -904,6 +905,7 @@ func InitConfig(config Config) {
 
 	setAssetFs(config)
 	setupAPM(config)
+	appsecconfig.InitConfig(config)
```
I think that for the sake of consistency, all of the "set known" and "bind env" stuff should be in this package, as it is for all other agents. Perhaps you can have a `setupAppsec` function in a separate file which can have your team as the owner.

Julio-Guerra replied:
Fixed: move `pkg/appsec/config.InitConfig()` to `pkg/config.setupAppSec()` in a new `appsec.go` file (added to CODEOWNERS too)

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -69,6 +71,11 @@ func NewAgent(ctx context.Context, conf *config.AgentConfig) *Agent {
 	in := make(chan *api.Payload, 1000)
 	statsChan := make(chan pb.StatsPayload, 100)
+	// Instantiate AppSec
+	appsecHandler, err := appsec.New(coreconfig.Datadog)
```
A few notes here, I'm adding numbers in case you want to answer each separately:

1. `coreconfig.Datadog` is a global object accessible from any package, why do you need to have it as an argument? I suspect you might've done it for tests, in which case, I'd say to make the public API `appsec.New` and have that call an internal unexported function `new` which takes the config and then use that for tests. No reason to sacrifice "public" API for testability.
2. I think that this code can safely go into the `api.NewHTTPReceiver` function. This seems like the wrong place.

Julio-Guerra replied:
Done as stated. I'll create the private `new` function if needed when writing the tests.

gbbr commented on pkg/trace/api/endpoints.go:
```diff
@@ -80,4 +83,20 @@ var endpoints = []endpoint{
+		Pattern: "/appsec",
+		Handler: func(r *HTTPReceiver) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				req.URL.Path = strings.TrimPrefix(req.URL.Path, "/appsec")
```
Why did you do this? Did you mean to be using https://golang.org/pkg/net/http/#StripPrefix instead?

Julio-Guerra replied:
I tried `http.StripPrefix` indeed, but it doesn't work from here: we end up either with 301 Moved Permanently with a request to `/appsec`; or 404 Not Found with a request to `/appsec/`.

gbbr replied:
Could it be that you might have been doing something wrong? Check https://play.golang.org/p/PqdIdFPWchG

Julio-Guerra replied:
Now fixed, but only with `/appsec/` instead of `/appsec` so that it gets stripped to `/` and the following AppSec ServeMux can properly match it.

For the record: the problem I had was coming from a mix if StripPrefix when it returns the empty string, and the AppSec ServeMux which wasn't happy about it because it expects "/" and not "".

gbbr commented on pkg/trace/agent/run.go:
```diff
@@ -149,6 +149,7 @@ func Run(ctx context.Context) {
 
 	agnt := NewAgent(ctx, cfg)
 	log.Infof("Trace agent running on host %s", cfg.Hostname)
+	// Run the tracer agent
```
Please remove this.

Julio-Guerra replied:
Done - glitch from the past version ^^'

Julio-Guerra committed:
appsec: refactor appsec configuration and instantiation

Julio-Guerra committed:
github: add appsec group to the codeowners list

Julio-Guerra committed:
trace/api: fix /appsec endpoint prefix stripping

Replace the /appsec endpoint by /appsec/.
The AppSec endpoint handler is an HTTP mux coming after the trace-agent mux.
So the trace-agent mux URL /appsec must be stripped before forwarding the
request to the AppSec mux so that it can properly find the next handler routes.

To do so, the trace-agent endpoint must be changed to /appsec/ so that it gets
stripped to / and we avoid the empty string which would lead to status
404 Not Found on the AppSec mux.

Note that by using http.StripPrefix(), a request to /appsec will be handled with
status code 301 Move Permanently to /appsec/

Julio-Guerra committed:
appsec/api: rename X-Api-Version into Api-Version

gbbr commented on pkg/config/appsec.go:
```diff
@@ -0,0 +1,9 @@
+
+// setupAppSec initializes the configuration values of the appsec agent.
+func setupAppSec(cfg Config) {
+	cfg.SetKnown("appsec_config.enabled")
+	cfg.SetKnown("appsec_config.appsec_dd_url")
```
The two options below automatically set things as "known", so you can remove these two lines.

Julio-Guerra replied:
Fixed 👍

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -87,6 +90,11 @@ func NewHTTPReceiver(conf *config.AgentConfig, dynConf *sampler.DynamicConfig, o
 	if features.Has("429") {
 		rateLimiterResponse = http.StatusTooManyRequests
 	}
+	// Instantiate the AppSec HTTP handler
```
I think the code is self-explanatory enough to not warrant this comment. Can we remove it please?

Julio-Guerra replied:
Removed 👍

Julio-Guerra committed:
appsec/api: move to an explicit URI-based API version

Julio-Guerra committed:
config/appsec: remove unecessary calls to SetKnown

Julio-Guerra committed:
trace/api: fix info endpoint test

Julio-Guerra committed:
trace/api: remove comment

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -28,6 +28,7 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/DataDog/datadog-agent/pkg/appsec"
```
This import is in the wrong group of imports.

Julio-Guerra replied:
Fixed

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -77,8 +78,10 @@ type HTTPReceiver struct {
 
-	wg   sync.WaitGroup // waits for all requests to be processed
-	exit chan struct{}
+	appsecHandler http.Handler
```
Can you move this field into the bigger group above please? The one with `out`, `conf`, etc.

Julio-Guerra replied:
Fixed

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -99,6 +106,8 @@ func NewHTTPReceiver(conf *config.AgentConfig, dynConf *sampler.DynamicConfig, o
 		debug:               strings.ToLower(conf.LogLevel) == "debug",
 		rateLimiterResponse: rateLimiterResponse,
 
+		appsecHandler: appsecHandler,
```
Same here.

Julio-Guerra replied:
Fixed

gbbr commented on pkg/appsec/config/config.go:
```diff
@@ -0,0 +1,68 @@
+)
+
+// AgentConfig represents the agent configuration API.
+type AgentConfig interface {
```
I recommend removing this. There are already mechanisms in place to facilitate testing the global configuration that you can use. No need to add extra layers. Furthermore, this isn't used now.

Julio-Guerra replied:
Done: replaced by the existing interface type `(pkg/config).Config`

gbbr commented on pkg/appsec/config/config.go:
```diff
@@ -0,0 +1,68 @@
+	// defaultIntakeURLTemplate specifies the string template allowing to obtain
+	// the intake URL from a given site.
+	defaultIntakeURLTemplate = "https://appsecevts-http-intake.logs.%s/v1/input"
+	// defaultSite is the intake site applied by default to
+	// defaultIntakeURLTemplate
```
I'd recommend documenting only _what it is_ not _where it is used_.

Julio-Guerra replied:
I think it does with "intake site applied by default to [the default intake url template] defaultIntakeURLTemplate.
I tried to make it more verbose. Let me know if anything else.

gbbr replied:
What I meant is:
```
// defaultSite is the default site used to target the intake.
```
Or similar. No need to mention the other constant, and how it is used. As far as we know, it may be used anywhere and it's not restricted to that purpose. But do as you please, I don't feel strongly here.

Julio-Guerra replied:
I see. Fixed.

gbbr commented on pkg/appsec/api/http/mux.go:
```diff
@@ -0,0 +1,18 @@
+// AppSec HTTP API. The API is versioned according to the X-Api-Version header
+// value so that requests are routed to the request multiplexer of the
+// given version.
+func NewServeMux(cfg *config.Config) *http.ServeMux {
```
This is overkill. You have a proxy inside a mux, inside another mux. There are no routes, there is just one proxy. Can we just drop everything and leave only the proxy? In `pkg/trace/endpoints.go` you can then link it straight to `/appsec/proxy`.

Julio-Guerra replied:
Done: now simplified to a single`appsec` package.

gbbr reviewed (changes requested):
* As far as `pkg/trace` goes, this is mostly fine, except some small nits.
* As for `pkg/appsec`, I think the current implementation is over engineered. All you need is one simple proxy handler and one single package. There is no need for multiple packages nor setting the ground for speculative future implementations. 

Can we please implement strictly what we need and not add code which is not needed today? That is: one handler and one package, which is a proxy, and that's it. What do you think?

gbbr reviewed (changes requested):
P.S. 

* You're going to also need to add metrics (probably best scoped as `datadog.trace_agent.appsec.*` which count incoming requests and their size. 
* You should also time the operation of forwarding the request using `timing.Since`. These will be of very high value for support tickets. 
* We also need to limit requests (you can start with 10MB which I assume should be plenty for a bunch of events - clients can comply).

You will have to add more customisation to the proxy to be able to achieve these...

Julio-Guerra committed:
appsec: refactoring into a single package

Julio-Guerra committed:
trace/api: fix info test

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -87,6 +89,10 @@ func NewHTTPReceiver(conf *config.AgentConfig, dynConf *sampler.DynamicConfig, o
 	}
+	appsecHandler, err := appsec.NewIntakeReverseProxy(conf.NewHTTPTransport())
+	if err != nil {
+		log.Error("appsec agent: ", err)
```
All log messages need to be capitalised.

Julio-Guerra replied:
Fixed

gbbr replied:
Not 🙂 

Also, there is no "appsec agent". Rather say:
```suggestion
		log.Errorf("Could not instantiate AppSec: %v", err)
```

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+// The reverse proxy handler also adds extra headers such as Dd-Api-Key and Via.
+func newIntakeReverseProxy(target *url.URL, apiKey string, transport http.RoundTripper) http.Handler {
+	proxy := httputil.NewSingleHostReverseProxy(target)
+	via := fmt.Sprintf("appsec-agent %s", info.Version)
```
```suggestion
	via := fmt.Sprintf("trace-agent %s", info.Version)
```

Julio-Guerra replied:
Fixed.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+		return disabled, errors.Wrap(err, "configuration: ")
+	}
+	if !cfg.Enabled {
+		return disabled, ErrAgentDisabled
```
I don't think this qualifies as an error messages. Disabling appsec is a potentially valid user intention.
```suggestion
		return disabled, nil
```

Julio-Guerra replied:
Ok, but I also added a log message before returning.

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -87,6 +89,10 @@ func NewHTTPReceiver(conf *config.AgentConfig, dynConf *sampler.DynamicConfig, o
 	if features.Has("429") {
 		rateLimiterResponse = http.StatusTooManyRequests
 	}
+	appsecHandler, err := appsec.NewIntakeReverseProxy(conf.NewHTTPTransport())
```
```suggestion
	appsecHandler, err := appsec.NewIntakeReverseProxy(conf.NewHTTPTransport())
```
👍🏻

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+
+func withMetrics(handler http.Handler) http.Handler {
+	const (
+		AppSecRequestMetricsPrefix = "datadog.trace_agent.appsec.api.request."
```
```suggestion
		AppSecRequestMetricsPrefix = "datadog.trace_agent.appsec."
```
We already have too many namespaces as is. No need to go that deep. We can then follow with:
* `.request`
* `.request_duration_ms`
* `.request_bytes`

This will match other metrics in convention

Julio-Guerra replied:
Fixed.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+func withMetrics(handler http.Handler) http.Handler {
+	const (
+		AppSecRequestMetricsPrefix = "datadog.trace_agent.appsec.api.request."
+		CountID                    = AppSecRequestMetricsPrefix + "count"
+		TimeID                     = AppSecRequestMetricsPrefix + "time"
+		ContentLengthID            = AppSecRequestMetricsPrefix + "content_length"
```
What's the point of having these constants? Can't we just inline them? They are only used once...

Julio-Guerra replied:
Just willing to avoid copy/paste mistakes + grouping all of them together to make it clear + relying on the compiler to do the concatenation at compilation time

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+
+		if cl := req.Header.Get("Content-Length"); cl != "" {
+			if cl, err := strconv.Atoi(cl); err == nil {
+				metrics.Histogram(ContentLengthID, float64(cl), tags, 1)
```
Nice 👍🏻

Julio-Guerra replied:
Now replaced by the actual body read size taken from the Count field of the LimitedReader struct

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+			}
+		}
+
+		now := time.Now()
+		defer func() {
```
Please use `timing.Since`

Julio-Guerra replied:
I'd love to, but I don't see how to pass it the tags.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,89 @@
+	}
+	proxy.Transport = transport
+	proxy.ErrorLog = stdlog.New(logutil.NewThrottled(5, 10*time.Second), "appsec backend proxy: ", 0)
+	return withMetrics(proxy)
```
This is a nice mechanism, but it fails if `Content-Length` is not set, and is inaccurate if for example the reader limit is reached. I'm afraid you won't be able to get away without writing the actual transport, as we do for profiles. Another requirement which asks for it is that with this approach, we're not able to create metrics for errors/failures.

Julio-Guerra replied:
Now that we rely on `LimitedReader`, this code was replaced with an access to its `Count` field.
As of the reached limit error, the `ErrorHandler` field of the reverse proxy does the job and allows to know when an error occurred during the `io.Copy` it does (tests will cover this case).

gbbr reviewed (commented):
Thanks Julio. This is much better in my opinion.

Julio-Guerra committed:
appsec: limit and monitor the request body size

Limit appsec request body reads to 5MB and monitor when is the limit
reached with a counter.

The body read limitation is done by `(trace/api).LimitedReader`, now
moved to a new `apiutil` package so that it can be imported from
`appsec` without import cycle.

Julio-Guerra committed:
trace/api: fix go import

Julio-Guerra committed:
appsec/cnfig: fix comment

Julio-Guerra committed:
appsec: log disabled state

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,115 @@
+	"github.com/DataDog/datadog-agent/pkg/trace/logutil"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/pkg/errors"
```
This import needs to be in its own group (external import). Just needs an empty line above it.

Julio-Guerra replied:
Fixed

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,115 @@
+	"github.com/pkg/errors"
+)
+
+// ErrAgentDisabled is the error message logged when the AppSec agent is
```
This is no longer used.

Julio-Guerra replied:
Indeed. Fixed.

gbbr reviewed (commented):
Cool. I think we can move on to adding tests. Besides unit tests, you might find it valuable to potentially extend the `pkg/trace/test` package to add a fake backend intake for appsec and ensure black box E2E testing. 

As for unit tests, please make sure to cover all ground, including all error branches, returned responses and generated metrics (see `(pkg/trace/test/testutil).TestStatsClient`)

Julio-Guerra committed:
appsec,trace: fix import groups

Julio-Guerra committed:
appsec: remove unecessary public error var

Julio-Guerra committed:
appsec: send internal proxy errors as json

Julio-Guerra committed:
config,appsec: add configuration tests

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,114 @@
+	disabled := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusNotImplemented)
+		json.NewEncoder(w).Encode("appsec api disabled")
```
I think you can just use `http.Error` here.

Julio-Guerra replied:
The intake api spec tells it returns a json value, even in case of error (explaining why), so I was thinking in aligning here.

nizox commented on pkg/trace/api/endpoints.go:
```diff
@@ -80,4 +80,8 @@ var endpoints = []endpoint{
 		Handler: func(r *HTTPReceiver) http.Handler { return http.HandlerFunc(r.handleStats) },
 	},
+	{
+		Pattern: "/appsec/v0.1/proxy/",
```
Can we use `/appsec/v1/input` instead?

In fact, as the agent is only a proxy for now, the `v1/input` suffix defines the "event sending protocol" while the base URL `http://localhost:8126/appsec/` & `https://appsecevts-http-intake.logs.datad0g.com/` are destinations.

The libraries could then send events to the agent or intake directly only by changing the base URL, which is useful for development and makes implementation more straightforward.

Julio-Guerra replied:
I'd rather have `/appsec/v0.1/proxy` forwarding to `https://appsecevts-http-intake.logs.datad0g.com` so that the agent API endpoint to use becomes `/appsec/v0.1/proxy/v1/input`

wdyt @gbbr? the base problem comes from the case when a library is in agentless mode. This would simplify a bit the configuration by just having to change the API URL "prefix" part (eg. http://localhost:8126/appsec/v0.1/proxy/v1/input would become https://appsecevts-http-intake.logs.datad0g.com/v1/input)

gbbr replied:
I think `/appsec/v1/input` is the right way to go. We use the same for profiling: `/profiling/v1/input`.

Julio-Guerra replied:
by the way, this is why the profiling api endpoint of the trace-agent is `/profiling/v1/input`: to have the same endpoint URL suffix as the intake's

gbbr replied:
I'd also be ok with `/appsec/proxy/v1/input`. That would still hold for the example given by @nizox. I think having a double version in one single URL is a bit much.

Julio-Guerra replied:
Done: `/appsec/v0.1/proxy` becomes `/appsec/v1/input`
The v0.1 Agent API versioning is dismissed for now. We'll reintroduce it in the future if needed by an actual agent API.

Julio-Guerra replied:
I think we are good to go with this during for the preview version.
There won't be any extra appsec endpoint as far as anyone could tell me.

Julio-Guerra committed:
appsec: add metrics tests

Julio-Guerra committed:
appsec: lower the log level to info when appsec is disabled

Julio-Guerra committed:
trace/api: align the appsec endpoint with the intake endpoint url

Ease library configurations by sharing the same endpoint suffix when
using the agent or directly the intake API.

Julio-Guerra committed:
appsec: add reverse proxy tests

albertvaka reviewed (approved):
👍 changes to codeowners

Julio-Guerra committed:
appsec: use ioutil.ReadAll to compile on older go versions

Julio-Guerra committed:
appsec: fix golint error

Julio-Guerra committed:
appsec: fix compilation error

Julio-Guerra committed:
appsec: fix spelling mistake

Julio-Guerra committed:
appsec: fix errcheck errors

Julio-Guerra committed:
appsec: remove flaky test on windows

Julio-Guerra committed:
appsec: return a reason when appsec is disabled

nizox reviewed (approved):
Tested E2E, all good.

Julio-Guerra committed:
appsec: set the default appsec site to datadoghq.com

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,128 @@
+	// Error metrics through the reverse proxy error handler
+	errorHandler := proxy.ErrorHandler
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		if err == apiutil.ErrLimitedReaderLimitReached {
```
We should count metrics for _all_ errors, not just for this one. A tag should specify which kind of error it is.

Julio-Guerra replied:
I did a change to monitor every error.
But getting the type of the error to distinguish them is sometimes too weak, for example when the error is created with `errors.New()` and the type doesn't describe the error.
Cf.  https://github.com/DataDog/datadog-agent/pull/8359/files#diff-204368e0d1216a43867fe43ec306add180e68d2b9db3414c95c0c08fd4d46ff0R90-R95

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,128 @@
+	errorHandler := proxy.ErrorHandler
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		if err == apiutil.ErrLimitedReaderLimitReached {
+			if err := metrics.Count(appSecRequestPayloadTooLargeMetricsID, 1, metricsTags(req), 1); err != nil {
```
Frankly we never check the error return value for this, anywhere 🙂

Julio-Guerra replied:
Ok, so we need to adjust errcheck in .golangci file tell not to fail on these calls, like it's done for `log.Error` =)

Julio-Guerra replied:
I changed `.golangci` so that `errcheck` doesn't trigger for calls to functions from `pkg/trace/metrics`.
And I simplified back the code by removing the error checks.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,128 @@
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		tags = append(tags, "content_type:"+ct)
+	}
+	if lr, ok := req.Body.(*apiutil.LimitedReader); ok {
```
The cardinality for this tag is too high. You'll have to remove it. It's already counted as a metric anyway.

Julio-Guerra replied:
Fixed

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,128 @@
+func withMetrics(proxy *httputil.ReverseProxy) http.Handler {
+	// Error metrics through the reverse proxy error handler
+	errorHandler := proxy.ErrorHandler
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
```
You could just do this above, where you define the ErrorHandler, but I'll leave it up to you...

Julio-Guerra replied:
Yes, just for the sake of having all the metrics in `withMetrics` :-)

gbbr replied:
Seems like a lot of hassle for little gain. If you would re-write the proxy transport you could have it all in one place. I don't feel strongly about it, as long as it works!

Julio-Guerra replied:
You're right, a custom RoundTripper is indeed clearer and cleaner 👍
Now fixed.

Julio-Guerra committed:
trace/api: fix error log message

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

Julio-Guerra committed:
appsec: metrics and error enhancements

Julio-Guerra committed:
appsec: fix error metrics

Julio-Guerra committed:
appsec: fix request error metrics

Move the request metrics from the proxy director and error handler to a proxy RoundTripper
wrapper. The proxy director was indeed copying the requests and making some parts of the
error monitoring fail because it wasn't using an older request value.

Julio-Guerra committed:
trace/api: change the appsec proxy endpoint to /appsec/proxy/

Allow both v1 and v2 API URLs by serving the reverse proxy endpoint /appsec/proxy/
as a prefix. Anything following is used in the intake api request.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,133 @@
+	defer func() {
+		tags := metricsTags(req)
+		if lr != nil {
+			metrics.Histogram(appSecRequestPayloadSizeMetricsID, float64(lr.Count), tags, 1)
```
This scares me a little. A race condition (panic) is possible here. See: https://github.com/golang/go/blob/go1.16/src/net/http/client.go#L134-L138.

We have 2 options:
* Alter the `LimitedReader` struct to access `Count` under a lock (and then retrieve it using a method `Count()`)
* Give up on this metric entirely

Julio-Guerra replied:
Removed as we couldn't find a proper way of making LimitedReader atomic.

gbbr reviewed (commented):
Almost there, one last concern :)

Julio-Guerra committed:
appsec: remove payload size metrics to avoid a race condition

Avoid a race condition on `LimitedReader.Count` with the body reader goroutine
which might be still reading the body even once the call to the RoundTripper
returns.

Note that making LimitedReader atomic was considered not possible due to the
impossibility to predict how many bytes Read() will actually read. Therefore
making impossible to predict and update the new read limit.

gbbr commented on pkg/appsec/appsec.go:
```diff
@@ -0,0 +1,127 @@
+		return disabled(fmt.Sprintf("appsec agent disabled due to a configuration error: %v", err)), errors.Wrap(err, "configuration: ")
+	}
+	if !cfg.Enabled {
+		log.Info("AppSec agent disabled. Set the environment variable `DD_APPSEC_ENABLED=true` or add the entry " +
```
This sounds like we're telling the user that something is wrong and that they should take action. I think the message: "AppSec proxy disabled." should be enough information. They already know the settings since they clearly manually set it to `false` (being it enabled by default).

Julio-Guerra replied:
Good point - fixed

gbbr commented on pkg/appsec/config.go:
```diff
@@ -0,0 +1,63 @@
+		Enabled:        cfg.GetBool("appsec_config.enabled"),
+		IntakeURL:      intakeURL,
+		APIKey:         cfg.GetString("api_key"),
+		MaxPayloadSize: cfg.GetInt64("appsec_config.max_payload_size"),
```
Is the value of `0` acceptable? If not, what should it mean? Should it mean "unlimited"? Probably not a good idea. I'd say let's set this only if its `> 0`. WDYT?

Julio-Guerra replied:
Yes, it was unlimited if <= 0
I changed it with a default payload size of 5 MB when <= 0

Julio-Guerra committed:
appsec: always limit the payload size

gbbr reviewed (approved):
🎉 Awesome! Thanks Julio!

P.S. No merging yet please, we are in code freeze 🥶

Julio-Guerra committed:
releasenote: add appsec proxy notes

Julio-Guerra committed:
Merge remote-tracking branch 'origin/main' into appsec-agent-beta

Julio-Guerra committed:
appsec: use 405 Not Allowed status code when disabled

Julio-Guerra committed:
appsec: use 405 Not Allowed status code when disabled

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -899,6 +899,32 @@ api_key:
+
+## @param appsec_config - custom object - optional
+## Enter specific configurations for your AppSec collection.
+## Uncomment this parameter and the one below to enable them.
```
```suggestion
## Uncomment the parameters `appsec_config` and `enabled` to enable AppSec.
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -899,6 +899,32 @@ api_key:
+# appsec_config:
+
+  ## @param enabled - boolean - optional - default: true
+  ## Set to true to enable the AppSec proxy in the APM Agent.
```
```suggestion
  ## Enable the AppSec proxy in the APM Agent.
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -899,6 +899,32 @@ api_key:
+  # enabled: true
+
+  ## @param appsec_dd_url - string - optional
+  ## Define the endpoint and port to hit when using a proxy for AppSec. The logs are forwarded in TCP
```
```suggestion
  ## Define the endpoint and port to hit when using a proxy for AppSec. Logs are forwarded by TCP;
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -899,6 +899,32 @@ api_key:
+
+  ## @param appsec_dd_url - string - optional
+  ## Define the endpoint and port to hit when using a proxy for AppSec. The logs are forwarded in TCP
+  ## therefore the proxy must be able to handle TCP connections.
```
```suggestion
  ## therefore, the proxy must be able to handle TCP connections.
```

ruthnaebeck commented on pkg/config/config_template.yaml:
```diff
@@ -899,6 +899,32 @@ api_key:
+  # appsec_dd_url: <ENDPOINT>:<PORT>
+
+  ## @param max_payload_size - integer - optional - default: 5242880 (5MB)
+  ## Define the maximum HTTP payload size (aka. the body) in bytes the AppSec proxy allows.
```
```suggestion
  ## Define the maximum HTTP payload size (the body) in bytes the AppSec proxy allows.
```

ruthnaebeck commented on releasenotes/notes/appsec-proxy-8e34c687cfc23e51.yaml:
```diff
@@ -0,0 +1,11 @@
+# Each section from every release note are combined when the
```
```suggestion
# Each section from every release note is combined when the
```

ruthnaebeck reviewed (changes requested):
Docs review

Julio-Guerra committed:
appsec: fix technical documentation

Co-authored-by: ruthnaebeck <19349244+ruthnaebeck@users.noreply.github.com>

ruthnaebeck reviewed (approved):
👍  for docs
